// ng-marquee/bundles/ng-marquee.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('ng-marquee', ['exports', '@angular/core'], factory) :
    (global = global || self, factory(global['ng-marquee'] = {}, global.ng.core));
}(this, (function (exports, core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/ng-marquee/ng-marquee.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgMarqueeComponent = /** @class */ (function () {
        function NgMarqueeComponent() {
        }
        /**
         * @private
         * @return {?}
         */
        NgMarqueeComponent.prototype.updateDuration = /**
         * @private
         * @return {?}
         */
        function () {
            if (!this.duration) {
                return;
            }
            this.animationElem.nativeElement.style.animationDuration = this.duration + "s";
        };
        /**
         * @return {?}
         */
        NgMarqueeComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
        };
        /**
         * @return {?}
         */
        NgMarqueeComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            this.updateDuration();
        };
        NgMarqueeComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng-marquee',
                        template: "<div class=\"ng-marquee-wrapper\">\n  <div #marqueeContainer>\n      <ng-content></ng-content>\n  </div>\n</div>",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        styles: [".ng-marquee-wrapper{overflow:hidden;position:relative}.ng-marquee-wrapper>div{height:100%;display:inline-block;margin:0;padding-left:100%;-webkit-animation:20s linear infinite marqueeAnimation;animation:20s linear infinite marqueeAnimation;white-space:nowrap}.ng-marquee-wrapper:hover>div{-webkit-animation-play-state:paused;animation-play-state:paused}@-webkit-keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@media (max-width:500px){.ng-marquee-wrapper>div{width:100%;height:100%;margin:0;transform:translateX(100%);-webkit-animation:9s linear 1s infinite marqueeAnimation;animation:9s linear 1s infinite marqueeAnimation;white-space:nowrap}}"]
                    }] }
        ];
        /** @nocollapse */
        NgMarqueeComponent.ctorParameters = function () { return []; };
        NgMarqueeComponent.propDecorators = {
            duration: [{ type: core.Input }],
            animationElem: [{ type: core.ViewChild, args: ['marqueeContainer', { static: true },] }]
        };
        return NgMarqueeComponent;
    }());
    if (false) {
        /** @type {?} */
        NgMarqueeComponent.prototype.duration;
        /** @type {?} */
        NgMarqueeComponent.prototype.animationElem;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/ng-marquee.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgMarqueeModule = /** @class */ (function () {
        function NgMarqueeModule() {
        }
        NgMarqueeModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [NgMarqueeComponent],
                        exports: [NgMarqueeComponent]
                    },] }
        ];
        return NgMarqueeModule;
    }());

    exports.NgMarqueeComponent = NgMarqueeComponent;
    exports.NgMarqueeModule = NgMarqueeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
