// @iplab/ngx-file-upload/esm5/lib/directives/attribute.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Optional, Host, Self } from '@angular/core';
import { FileUploadComponent } from '../components/multiple-file-upload/file-upload.component';
import { SimpleFileUploadComponent } from '../components/simple-file-upload/simple-file-upload.component';
var FilesAcceptDirective = /** @class */ (function () {
    function FilesAcceptDirective(fileUpload, simpleFileUpload) {
        this.fileUpload = null;
        this.fileUpload = fileUpload || simpleFileUpload;
    }
    /**
     * @return {?}
     */
    FilesAcceptDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.setAccept(this.accept);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FilesAcceptDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('accept' in changes && changes['accept'].currentValue !== changes['accept'].previousValue) {
            this.setAccept(this.accept);
        }
    };
    /**
     * @private
     * @param {?} accept
     * @return {?}
     */
    FilesAcceptDirective.prototype.setAccept = /**
     * @private
     * @param {?} accept
     * @return {?}
     */
    function (accept) {
        if (this.fileUpload && this.fileUpload.control) {
            this.fileUpload.control.acceptFiles(accept);
        }
    };
    FilesAcceptDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'file-upload[accept]',
                    host: { '[attr.accept]': 'accept ? accept : null' }
                },] }
    ];
    /** @nocollapse */
    FilesAcceptDirective.ctorParameters = function () { return [
        { type: FileUploadComponent, decorators: [{ type: Optional }, { type: Host }, { type: Self }] },
        { type: SimpleFileUploadComponent, decorators: [{ type: Optional }, { type: Host }, { type: Self }] }
    ]; };
    FilesAcceptDirective.propDecorators = {
        accept: [{ type: Input }]
    };
    return FilesAcceptDirective;
}());
export { FilesAcceptDirective };
if (false) {
    /** @type {?} */
    FilesAcceptDirective.prototype.accept;
    /**
     * @type {?}
     * @private
     */
    FilesAcceptDirective.prototype.fileUpload;
}
