// ng-marquee/esm5/lib/ng-marquee/ng-marquee.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-marquee/ng-marquee.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, ViewChild, ElementRef } from '@angular/core';
var NgMarqueeComponent = /** @class */ (function () {
    function NgMarqueeComponent() {
    }
    /**
     * @private
     * @return {?}
     */
    NgMarqueeComponent.prototype.updateDuration = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.duration) {
            return;
        }
        this.animationElem.nativeElement.style.animationDuration = this.duration + "s";
    };
    /**
     * @return {?}
     */
    NgMarqueeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    NgMarqueeComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.updateDuration();
    };
    NgMarqueeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-marquee',
                    template: "<div class=\"ng-marquee-wrapper\">\n  <div #marqueeContainer>\n      <ng-content></ng-content>\n  </div>\n</div>",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".ng-marquee-wrapper{overflow:hidden;position:relative}.ng-marquee-wrapper>div{height:100%;display:inline-block;margin:0;padding-left:100%;-webkit-animation:20s linear infinite marqueeAnimation;animation:20s linear infinite marqueeAnimation;white-space:nowrap}.ng-marquee-wrapper:hover>div{-webkit-animation-play-state:paused;animation-play-state:paused}@-webkit-keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@media (max-width:500px){.ng-marquee-wrapper>div{width:100%;height:100%;margin:0;transform:translateX(100%);-webkit-animation:9s linear 1s infinite marqueeAnimation;animation:9s linear 1s infinite marqueeAnimation;white-space:nowrap}}"]
                }] }
    ];
    /** @nocollapse */
    NgMarqueeComponent.ctorParameters = function () { return []; };
    NgMarqueeComponent.propDecorators = {
        duration: [{ type: Input }],
        animationElem: [{ type: ViewChild, args: ['marqueeContainer', { static: true },] }]
    };
    return NgMarqueeComponent;
}());
export { NgMarqueeComponent };
if (false) {
    /** @type {?} */
    NgMarqueeComponent.prototype.duration;
    /** @type {?} */
    NgMarqueeComponent.prototype.animationElem;
}
