// ng-marquee/fesm2015/ng-marquee.js
import { Component, ChangeDetectionStrategy, Input, ViewChild, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-marquee/ng-marquee.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgMarqueeComponent {
    constructor() { }
    /**
     * @private
     * @return {?}
     */
    updateDuration() {
        if (!this.duration) {
            return;
        }
        this.animationElem.nativeElement.style.animationDuration = `${this.duration}s`;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.updateDuration();
    }
}
NgMarqueeComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-marquee',
                template: "<div class=\"ng-marquee-wrapper\">\n  <div #marqueeContainer>\n      <ng-content></ng-content>\n  </div>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".ng-marquee-wrapper{overflow:hidden;position:relative}.ng-marquee-wrapper>div{height:100%;display:inline-block;margin:0;padding-left:100%;-webkit-animation:20s linear infinite marqueeAnimation;animation:20s linear infinite marqueeAnimation;white-space:nowrap}.ng-marquee-wrapper:hover>div{-webkit-animation-play-state:paused;animation-play-state:paused}@-webkit-keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@keyframes marqueeAnimation{0%{transform:translateX(0)}100%{transform:translateX(-100%)}}@media (max-width:500px){.ng-marquee-wrapper>div{width:100%;height:100%;margin:0;transform:translateX(100%);-webkit-animation:9s linear 1s infinite marqueeAnimation;animation:9s linear 1s infinite marqueeAnimation;white-space:nowrap}}"]
            }] }
];
/** @nocollapse */
NgMarqueeComponent.ctorParameters = () => [];
NgMarqueeComponent.propDecorators = {
    duration: [{ type: Input }],
    animationElem: [{ type: ViewChild, args: ['marqueeContainer', { static: true },] }]
};
if (false) {
    /** @type {?} */
    NgMarqueeComponent.prototype.duration;
    /** @type {?} */
    NgMarqueeComponent.prototype.animationElem;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-marquee.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgMarqueeModule {
}
NgMarqueeModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgMarqueeComponent],
                exports: [NgMarqueeComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ng-marquee.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NgMarqueeComponent, NgMarqueeModule };
